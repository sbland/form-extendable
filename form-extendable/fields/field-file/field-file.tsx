// TODO: The types in this file need checking
import React, { useState, useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { FileManager } from '@react_db_client/components.file-manager';
import { ItemList } from '@react_db_client/components.item-list';
import {
  IFieldComponentProps,
  IFile,
  IHeadingFile,
  IPopupProps,
} from '@form-extendable/lib';
import { FilterObjectClass } from '@react_db_client/constants.client-types';

export interface IFieldFileProps<V extends IFile | IFile[]>
  extends IFieldComponentProps<V>,
    IHeadingFile<V> {
  fileServerUrl: string;
  asyncGetDocuments: (filters: FilterObjectClass[]) => Promise<IFile[]>;
  PopupPanel: React.FC<IPopupProps>;
}

export type Foo = IFieldFileProps<IFile>['onChange'];

/**
 * Form component file field
 *
 * Value should be either a single or array of file objects
 *
 */
export const FieldFile: React.FC<IFieldFileProps<IFile | IFile[]>> = ({
  uid,
  multiple,
  onChange,
  collectionId,
  documentId,
  fileType,
  value,
  fileServerUrl,
  asyncGetDocuments,
  // TODO: Add required check
  // required,
  PopupPanel,
}) => {
  if (
    value &&
    (typeof value !== 'object' || (value[0] && typeof value[0] !== 'object'))
  )
    throw Error(`Value must be file type. Got ${value}`);

  const [fileList, setFileList] = useState<IFile[]>(() =>
    value && Array.isArray(value) ? value : (value && [value]) || []
  );
  const [showFileSelectionPanel, setShowFileSelectionPanel] = useState(false);

  useEffect(() => {
    setFileList(
      value && Array.isArray(value) ? value : (value && [value]) || []
    );
  }, [value]);

  const handleSelected = (file: IFile, fileData: IFile | IFile[]) => {
    const newFileList = multiple
      ? [...fileList, ...(fileData as IFile[])]
      : [fileData as IFile];
    setFileList(newFileList);
    setShowFileSelectionPanel(false);
    const newData = multiple
      ? [...fileList, ...(fileData as IFile[])]
      : (fileData as IFile);
    onChange && onChange(newData as typeof value);
  };

  const handleFileDelete = (fuid) => {
    const newFileList = fileList.filter((f) => f.uid !== fuid);
    setFileList(newFileList);
    const newData = multiple ? newFileList : null;
    onChange && onChange(newData);
  };

  const filesData = useMemo(
    () =>
      fileList &&
      fileList
        .filter((f) => f)
        .map((file) => ({
          uid: file.uid || file.name,
          label: file.name,
          type: fileType === 'image' ? 'image' : 'button',
          src: `${fileServerUrl}/${file.filePath}`,
        })),
    [fileList, fileType]
  );

  return (
    <>
      <PopupPanel
        isOpen={showFileSelectionPanel}
        handleClose={() => setShowFileSelectionPanel(false)}
        title="File Selector"
      >
        <FileManager
          handleSelect={handleSelected}
          collectionId={collectionId}
          documentId={documentId}
          fileType={fileType}
          allowMultiple={multiple}
          asyncGetDocuments={asyncGetDocuments}
          fileServerUrl={fileServerUrl}
        />
      </PopupPanel>
      <div className="FieldFile">
        <div>
          <button
            type="button"
            className="button-one addFileBtn"
            onClick={() => setShowFileSelectionPanel(true)}
          >
            Add File
          </button>
        </div>
        <div style={{ display: 'flex' }}>
          Files:
          <ItemList
            viewType={fileType === 'image' ? 'grid' : 'list'}
            items={filesData}
            overlayButtons={[
              {
                uid: 'remove',
                func: (fuid) => handleFileDelete(fuid),
                label: 'Remove',
                icon: '🗑️',
              },
            ]}
          />
        </div>
      </div>
    </>
  );
};

const fileValueShape = {
  filePath: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
};

FieldFile.propTypes = {
  uid: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.shape(fileValueShape),
    PropTypes.arrayOf(PropTypes.shape(fileValueShape)),
  ]),
  onChange: PropTypes.func.isRequired,
  collectionId: PropTypes.string.isRequired,
  documentId: PropTypes.string.isRequired,
  fileType: PropTypes.oneOf(['image', 'document', 'data', '*']),
  multiple: PropTypes.bool,
  fileServerUrl: PropTypes.string.isRequired,
  asyncGetDocuments: PropTypes.func.isRequired,
  PopupPanel: PropTypes.elementType.isRequired,
};

FieldFile.defaultProps = {
  value: [],
  multiple: false,
  fileType: '*',
  PopupPanel: (({ children, isOpen }) =>
    isOpen ? children : '') as React.FC<IPopupProps>,
};
