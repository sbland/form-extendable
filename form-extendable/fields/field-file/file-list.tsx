import React from 'react';
import { EFileType, IFile, Uid } from '@react_db_client/constants.client-types';
import {
  FileItemBtnWrap,
  FileItemDeleteBtn,
  FileListImageItemStyle,
  FileListItemStyle,
  ImageItemDeleteBtn,
} from './styles';
import { Emoji } from '@react_db_client/components.emoji';

export interface IFileItem {
  fileData: IFile;
  fileType: EFileType;
  fileServerUrl: string;
  deleteFile: (fileId: Uid) => void;
}

export const FileItem = ({
  fileData,
  fileType,
  fileServerUrl,
  deleteFile,
}: IFileItem) => {
  const fullFilePath = `${fileServerUrl}/${[fileData.filePath, fileData.name]
    .filter((f) => f)
    .join('/')}`;
  switch (fileType as EFileType) {
    case EFileType.IMAGE:
      return (
        <FileListImageItemStyle key={fileData.uid}>
          <img src={fullFilePath} alt={fileData.label} role="img" />
          <FileItemBtnWrap>
            <ImageItemDeleteBtn onClick={() => deleteFile(fileData.uid)}>
              <Emoji emoj="🗑️" label="Remove" />
            </ImageItemDeleteBtn>
          </FileItemBtnWrap>
        </FileListImageItemStyle>
      );
    default:
      return (
        <FileListItemStyle key={fileData.uid}>
          <p data-testid={`fileLabel_${fileData.uid}`}>{fileData.label}</p>
          <FileItemBtnWrap>
            <FileItemDeleteBtn onClick={() => deleteFile(fileData.uid)}>
              <Emoji emoj="🗑️" label="Remove" />
            </FileItemDeleteBtn>
          </FileItemBtnWrap>
        </FileListItemStyle>
      );
  }
};

export interface IFilesListProps {
  items: IFile[];
  fileType: EFileType;
  fileServerUrl: string;
  deleteFile: (fileId: Uid) => void;
}
