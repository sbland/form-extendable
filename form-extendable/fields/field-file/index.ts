export { FieldFile } from './field-file';
export { FieldFileMultiple } from './field-file-multiple';
export { getDisplayValue, editValue } from './test-utils';
export type { THeadingTypes } from './test-utils';
export type { IFieldFileProps } from './field-file';
export type { IFieldFileMultipleProps } from './field-file-multiple';
