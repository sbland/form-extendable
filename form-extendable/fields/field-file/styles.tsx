import styled from 'styled-components';

export const FileListStyle = styled.ul`
  list-style: none;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  margin: 0;
  padding: 0;
  justify-content: flex-start;
  flex-wrap: wrap;
  align-items: flex-end;
`;

export const FileListItemStyle = styled.li`
  position: relative;
  border: 1px solid grey;
  border-radius: 1rem;
  overflow: hidden;
  margin: 0;
  padding: 0;
  max-height: 10rem;
  min-height: 1.6rem;
  padding: 0 0 0 0.4rem;
  margin-left: 0.5rem;
  margin-top: 0.5rem;
  line-height: 1.6rem;
  min-width: 3rem;

  display: flex;

  img {
    height: 100%;
    max-width: 100%;
    margin-right: 0.2rem;
  }
  p {
    margin: 0;
    padding: 0;
    margin-right: 0.2rem;
  }
`;

export const FileListImageItemStyle = styled(FileListItemStyle)`
  height: 10rem;
  padding: 0.4rem;
`;

export const FileItemDeleteBtn = styled.button`
  width: 1.6rem;
  height: 1.6rem;
  line-height: 2rem;
  background: white;
  border: 1px solid grey;
  border-radius: 2rem;
  cursor: pointer;
  margin: 0;
  padding: 0;
  span {
    margin: 0;
    padding: 0;
  }

  border: none;
  color: #555;
  outline: 1px solid #ddd;
  &:hover {
    background: #ddd;
    outline: none;
  }
`;

export const ImageItemDeleteBtn = styled(FileItemDeleteBtn)`
  width: 3rem;
  height: 3rem;
`;

export const FileItemBtnWrap = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const AddFileListItem = styled(FileListItemStyle)`
  border: none;
  display: wrap;
  flex-direction: column;
  justify-content: center;
  align-self: center;
`;

export const AddFileButton = styled.button`
  height: 3rem;
  width: 3rem;
  font-size: 3rem;
  line-height: 3rem;
  font-weight: 100;
  background: white;
  border-radius: 4rem;
  cursor: pointer;
  border: none;
  color: #555;
  outline: 1px solid #ddd;
  &:hover {
    background: #ddd;
    outline: none;
  }
`;
