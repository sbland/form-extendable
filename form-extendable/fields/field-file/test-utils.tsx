import React from 'react';
import { screen, within } from '@testing-library/react';
import UserEvent from '@testing-library/user-event';
import {
  IHeadingFile,
  IHeadingFileMulti,
  IHeadingImage,
  IHeadingImageMulti,
} from '@form-extendable/lib';
import {
  EFileType,
  EFilterType,
  IFile,
} from '@react_db_client/constants.client-types';

export type THeadingTypes =
  | IHeadingImage
  | IHeadingFile
  | IHeadingImageMulti
  | IHeadingFileMulti;

const promiseAllInSeries = async (iterable) => {
  for (const x of iterable) {
    await x();
  }
};

const queryFileAlreadyUploaded = async (
  value: IFile,
  heading: THeadingTypes,
  fileManager: HTMLElement
) => {
  const searchField = within(fileManager).getByPlaceholderText('search...');
  await UserEvent.click(searchField);
  await UserEvent.keyboard(value.label);

  const sasPanels = within(fileManager).getByTestId(
    'rdc-sas-file-manager-existing-files'
  );
  const existingFilesList = within(sasPanels).getByRole('list');

  const displayValue =
    heading.fileType === EFileType.IMAGE ? value.name : value.label;

  return within(existingFilesList).queryByText(displayValue);
};

export const editValueCommon = async (
  value: IFile,
  heading: THeadingTypes,
  fileManager: HTMLElement
) => {
  const selectionInExistingList = await queryFileAlreadyUploaded(
    value,
    heading,
    fileManager
  );

  if (!selectionInExistingList) {
    const selectFilesBtn = within(fileManager).getByLabelText('Select Files');
    if (!value.data)
      throw Error(`Must supply file data! Fieldid: ${heading.uid}`);
    await UserEvent.upload(selectFilesBtn, value.data);
    const uploadBtn = within(fileManager).getByRole('button', {
      name: 'Upload',
    });
    await UserEvent.click(uploadBtn);

    const sasPanels = within(fileManager).getByTestId(
      'rdc-sas-file-manager-existing-files'
    );
    const loadedFilesList = within(sasPanels).getByRole('list');

    await within(loadedFilesList).findByText(value.name);
  }
};

export const selectInUploadedList = async (
  value: IFile,
  fileManager: HTMLElement
) => {
  const sasPanels = within(fileManager).getByTestId(
    'rdc-sas-file-manager-existing-files'
  );
  const loadedFilesList = within(sasPanels).getByRole('list');

  await within(loadedFilesList).findByText(value.label);
  const selectItemBtn = within(loadedFilesList).getByRole('button', {
    name: value.label,
  });
  await UserEvent.click(selectItemBtn);
};

export const editValueSingle = async (
  value: IFile,
  formEl: HTMLFormElement,
  heading: THeadingTypes
) => {
  const fieldComponent = within(formEl).getByTestId(
    `${heading.type}-${heading.uid}`
  );

  const selectFileBtn: HTMLButtonElement = within(fieldComponent).getByRole(
    'button',
    {
      name: 'swap',
    }
  );
  await UserEvent.click(selectFileBtn);
  const fileManager = await screen.findByTestId('rdc-file-manager');
  await editValueCommon(value, heading, fileManager);
  await selectInUploadedList(value, fileManager);
};

export const editValueMulti = async (
  value: IFile[],
  formEl: HTMLFormElement,
  heading: THeadingTypes
) => {
  const fieldComponent = within(formEl).getByTestId(
    `${heading.type}-${heading.uid}`
  );

  const selectFileBtn: HTMLButtonElement = within(fieldComponent).getByRole(
    'button',
    {
      name: 'add',
    }
  );
  await UserEvent.click(selectFileBtn);
  const fileManager = await screen.findByTestId('rdc-file-manager');
  await promiseAllInSeries(
    value.map((v) => async () => {
      await editValueCommon(v, heading, fileManager);
    })
  );

  await promiseAllInSeries(
    value.map((v) => async () => {
      await selectInUploadedList(v, fileManager);
    })
  );

  const acceptSelectionBtn = within(fileManager).getByRole('button', {
    name: /Accept Selection/,
  });
  await UserEvent.click(acceptSelectionBtn);
};

export const editValue = async (
  value: IFile | IFile[],
  formEl: HTMLFormElement,
  heading: THeadingTypes
) => {
  if (heading.type === EFilterType.fileMultiple || heading.multiple) {
    return editValueMulti(value as IFile[], formEl, heading);
  } else {
    return editValueSingle(value as IFile, formEl, heading);
  }
};

export const getDisplayValue = async <T,>(
  formEl: HTMLFormElement,
  heading: THeadingTypes
): Promise<string> => {
  const fieldComponent = within(formEl).getByTestId(
    `${heading.type}-${heading.uid}`
  );
  if (
    (heading.type === EFilterType.file ||
      heading.type === EFilterType.fileMultiple) &&
    heading.fileType !== EFileType.IMAGE
  ) {
    const filesList = within(fieldComponent).getByTestId(
      `files-list-${heading.uid}`
    );

    const filesListItems: string[] = within(filesList)
      .getAllByRole('listitem')
      .map((f) => within(f).queryByTestId('fileLabel_', { exact: false }))
      .map((f) => f && f.textContent)
      .filter((f) => f != null) as string[];
    return filesListItems.filter((f) => f).join(',');
  } else if (
    heading.type === EFilterType.image ||
    heading.fileType === EFileType.IMAGE
  ) {
    const filesList = within(fieldComponent).getByTestId(
      `files-list-${heading.uid}`
    );
    const filesListItems: HTMLImageElement[] = within(filesList)
      .getAllByRole('listitem')
      .map<HTMLImageElement[]>((f) => within(f).queryAllByRole('img'))
      .map((f) => f && f[0])
      .filter((f) => f != null) as HTMLImageElement[];
    const imgAlts = filesListItems.filter((f) => f).map((f) => f.alt);
    return imgAlts.filter((f) => f).join(',');
  } else {
    console.info(heading);
    throw new Error(`Not Implemented: ${heading.uid}`);
  }
};
