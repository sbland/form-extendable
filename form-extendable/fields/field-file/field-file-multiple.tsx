import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { FileManager } from '@react_db_client/components.file-manager';
import { Emoji } from '@react_db_client/components.emoji';
import {
  EFileType,
  FilterObjectClass,
  IFile,
  Uid,
} from '@react_db_client/constants.client-types';
import {
  IFieldComponentProps,
  IHeadingFileMulti,
  IPopupProps,
} from '@form-extendable/lib';
import { FileItem } from './file-list';
import { AddFileButton, AddFileListItem, FileListStyle } from './styles';

export interface IFieldFileMultipleProps
  extends IFieldComponentProps<IFile[], IHeadingFileMulti> {
  fileServerUrl: string;
  asyncGetFiles: (
    metaData?: any
  ) => (filters?: FilterObjectClass[]) => Promise<IFile[]>;
  asyncFileUpload: (
    metaData?: any
  ) => (
    data: File,
    fileType: EFileType,
    callback: () => void,
    fileMetaData: Partial<IFile>
  ) => Promise<void>;
  PopupPanel: React.FC<IPopupProps>;
}

/**
 * Form component file field
 *
 * Value should be either a single or array of file objects
 *
 */
export const FieldFileMultiple: React.FC<IFieldFileMultipleProps> = ({
  uid,
  onChange,
  fileType,
  value: fileList,
  fileServerUrl,
  asyncGetFiles,
  asyncFileUpload,
  metaData,
  // TODO: Add required check
  // required,
  PopupPanel,
}) => {
  if (fileList && !Array.isArray(fileList))
    throw Error(
      `Value must be an array of file type. Got ${fileList} with type ${typeof fileList}`
    );

  const [showFileSelectionPanel, setShowFileSelectionPanel] = useState(false);

  const handleSelected = (fileData: IFile | IFile[] | null) => {
    if (!Array.isArray(fileData))
      throw Error('Must return fileData as array if multiple is true');

    setShowFileSelectionPanel(false);
    const newData = [...(fileList || []), ...(fileData as IFile[])];
    onChange && onChange(newData);
  };

  const handleFileDelete = (fuid) => {
    const newFileList = (fileList as IFile[]).filter((f) => f.uid !== fuid);
    onChange && onChange(newFileList);
  };

  const itemsRendered = React.useMemo(
    () =>
      fileList?.map((f) => (
        <FileItem
          key={f.uid}
          fileData={f}
          fileType={fileType}
          fileServerUrl={fileServerUrl}
          deleteFile={(fuid: Uid) => handleFileDelete(fuid)}
        />
      )),
    [fileList, fileType, fileServerUrl]
  );

  return (
    <>
      <PopupPanel
        id={`field-file-popup_${uid}`}
        isOpen={showFileSelectionPanel}
        onClose={() => setShowFileSelectionPanel(false)}
        title="File Selector"
      >
        <FileManager
          handleSelect={handleSelected}
          fileType={fileType}
          allowMultiple
          asyncGetFiles={asyncGetFiles(metaData)}
          fileServerUrl={fileServerUrl}
          asyncFileUpload={asyncFileUpload(metaData)}
        />
      </PopupPanel>
      <div className="FieldFile">
        <div style={{ display: 'flex' }} data-testid={`files-list-${uid}`}>
          <FileListStyle>
            <AddFileListItem>
              <AddFileButton
                type="button"
                className="button-one addFileBtn"
                onClick={() => setShowFileSelectionPanel(true)}
              >
                <Emoji emoj="âž•" label="add" />
              </AddFileButton>
            </AddFileListItem>
            {itemsRendered}
          </FileListStyle>
        </div>
      </div>
    </>
  );
};

const fileValueShape = {
  filePath: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  fileType: PropTypes.oneOf(Object.values(EFileType)).isRequired,
};

const valueValidator = PropTypes.shape(fileValueShape).isRequired;

const valueValidatorFull = PropTypes.oneOfType([
  valueValidator,
  PropTypes.arrayOf(valueValidator).isRequired,
]);

FieldFileMultiple.propTypes = {
  uid: PropTypes.string.isRequired,
  // @ts-ignore
  value: valueValidatorFull,
  onChange: PropTypes.func.isRequired,
  fileType: PropTypes.oneOf(Object.values(EFileType)).isRequired,
  // @ts-ignore
  multiple: PropTypes.bool,
  fileServerUrl: PropTypes.string.isRequired,
  asyncGetFiles: PropTypes.func.isRequired,
  asyncFileUpload: PropTypes.func.isRequired,
  PopupPanel: PropTypes.func.isRequired, // React.FC
};

FieldFileMultiple.defaultProps = {
  value: [],
  multiple: true,
  fileType: EFileType.ANY,
  PopupPanel: (({ children, isOpen }) =>
    isOpen ? children : '') as React.FC<IPopupProps>,
};
