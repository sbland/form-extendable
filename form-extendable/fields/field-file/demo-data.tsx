import React from 'react';
import { IPopupProps } from '@form-extendable/lib';
import {
  EFileType,
  EFilterType,
  IFile,
} from '@react_db_client/constants.client-types';
import { IFieldFileProps } from './field-file';
import { IFieldFileMultipleProps } from './field-file-multiple';

export const DEMO_IMAGE_FILES_DATA: IFile[] = [
  {
    uid: 'a',
    filePath: '',
    label: 'FileA',
    name: 'bit-logo.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'b',
    filePath: 'extensions-icons',
    label: 'FileB',
    name: 'vue_grey.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'c',
    filePath: 'Community/app-components',
    label: 'FileC',
    name: 'card.jpg',
    fileType: EFileType.IMAGE,
  },
];

export const DEMO_FILES_DATA: IFile[] = [
  {
    uid: 'a',
    filePath: '',
    label: 'FileA with a really long name',
    name: 'bit-logo.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'b',
    filePath: 'extensions-icons',
    label: 'FileB',
    name: 'vue_grey.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'c',
    filePath: 'Community/app-components',
    label: 'FileC',
    name: 'card.jpg',
    fileType: EFileType.IMAGE,
  },
];

export const DEMO_FILES_DATA_MANY = [
  ...DEMO_FILES_DATA,
  ...Array(100)
    .fill(0)
    .map((_, i) => ({
      uid: `file_${i}`,
      filePath: '',
      label: `FileA with a really long name ${i}`,
      name: 'bit-logo.svg',
      fileType: EFileType.IMAGE,
    })),
];

export const DEMO_IMAGE_FILES_MANY = [
  ...DEMO_IMAGE_FILES_DATA,
  ...Array(100)
    .fill(0)
    .map((_, i) => ({
      uid: `file_${i}`,
      filePath: '',
      label: `FileA with a really long name ${i}`,
      name: `bit-logo_${i}.svg`,
      fileType: EFileType.IMAGE,
    })),
];

export const PopupPanel: React.FC<IPopupProps> = ({
  children,
  isOpen,
  handleClose,
}) => (isOpen ? <>{children}</> : <></>);

const onChange: IFieldFileProps['onChange'] = () => {};
const onChangeMulti: IFieldFileMultipleProps['onChange'] = () => {};

export const dummyProps: IFieldFileProps = {
  uid: 'uid',
  label: 'File Field',
  type: EFilterType.file,
  multiple: false,
  onChange,
  fileType: EFileType.IMAGE,
  value: DEMO_IMAGE_FILES_DATA[0],
  fileServerUrl: 'https://static.bit.dev',
  asyncGetFiles: (metaData) => async () => DEMO_IMAGE_FILES_DATA,
  PopupPanel,
  asyncFileUpload: () => async () => {},
};

export const dummyPropsImagesMany: IFieldFileMultipleProps = {
  ...dummyProps,
  onChange: onChangeMulti,
  type: EFilterType.fileMultiple,
  defaultValue: [],
  multiple: true,
  value: [...DEMO_IMAGE_FILES_MANY].slice(0, 50),
  asyncGetFiles: (metaData) => async () =>
    [...DEMO_IMAGE_FILES_MANY].slice(50, 60),
};

export const dummyPropsDocs: IFieldFileProps = {
  ...dummyProps,
  fileType: EFileType.DOCUMENT,
  value: DEMO_FILES_DATA[0],
  asyncGetFiles: (metaData) => async () => DEMO_FILES_DATA,
};

export const dummyPropsDocsMany: IFieldFileMultipleProps = {
  ...dummyPropsDocs,
  onChange: onChangeMulti,
  type: EFilterType.fileMultiple,
  defaultValue: [],
  multiple: true,
  value: [...DEMO_FILES_DATA_MANY].slice(0, 50),
  asyncGetFiles: (metaData) => async () =>
    [...DEMO_FILES_DATA_MANY].slice(50, 60),
};
