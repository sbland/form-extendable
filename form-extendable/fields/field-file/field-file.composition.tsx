import React from 'react';
import { screen } from '@testing-library/react';
import { CompositionWrapDefault } from '@react_db_client/helpers.composition-wraps';
import { WrapFieldComponent } from '@form-extendable/composition-helpers';
import { FieldFile } from './field-file';
import {
  DEMO_FILES_DATA,
  DEMO_FILES_DATA_MANY,
  DEMO_IMAGE_FILES_DATA,
  DEMO_IMAGE_FILES_MANY,
  dummyProps,
  dummyPropsDocs,
  dummyPropsDocsMany,
  dummyPropsImagesMany,
} from './demo-data';

export const BasicFieldFileImages = () => (
  <CompositionWrapDefault height="40rem" width="60rem">
    <WrapFieldComponent>
      <FieldFile {...dummyProps} />
    </WrapFieldComponent>
  </CompositionWrapDefault>
);

BasicFieldFileImages.waitForReady = async () => {
  await screen.findAllByAltText(DEMO_IMAGE_FILES_DATA[0].label);
};

export const BasicFieldFileImagesMany = () => (
  <CompositionWrapDefault height="40rem" width="60rem">
    <WrapFieldComponent>
      <FieldFile {...dummyPropsImagesMany} />
    </WrapFieldComponent>
  </CompositionWrapDefault>
);

BasicFieldFileImagesMany.waitForReady = async () => {
  await screen.findAllByAltText(DEMO_IMAGE_FILES_MANY[0].label);
};

export const BasicFieldFileDocs = () => (
  <CompositionWrapDefault height="40rem" width="60rem">
    <WrapFieldComponent>
      <FieldFile {...dummyPropsDocs} />
    </WrapFieldComponent>
  </CompositionWrapDefault>
);

BasicFieldFileDocs.waitForReady = async () => {
  await screen.findAllByText(DEMO_FILES_DATA[0].label);
};

export const BasicFieldFileDocsmany = () => (
  <CompositionWrapDefault height="40rem" width="60rem">
    <WrapFieldComponent>
      <FieldFile {...dummyPropsDocsMany} />
    </WrapFieldComponent>
  </CompositionWrapDefault>
);

BasicFieldFileDocsmany.waitForReady = async () => {
  await screen.findAllByText(DEMO_FILES_DATA_MANY[0].label);
};
