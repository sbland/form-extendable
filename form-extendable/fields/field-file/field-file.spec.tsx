import React from 'react';
import UserEvent from '@testing-library/user-event';
import { FieldFile, IFieldFileProps, IPopupProps } from './field-file';
import * as compositions from './field-file.composition';
import { DEMO_FILES_DATA } from './demo-data';
import { render, screen } from '@testing-library/react';

const onChange = jest.fn();
const asyncGetDocuments = jest.fn();
const PopupPanel: React.FC<IPopupProps> = ({
  children,
  isOpen,
  handleClose,
}) => <>{children}</>;

const defaultProps: IFieldFileProps = {
  uid: 'uid',
  multiple: false,
  onChange,
  collectionId: 'collectionId',
  documentId: 'documentId',
  fileType: 'image',
  value: DEMO_FILES_DATA,
  fileServerUrl: 'fileserver',
  asyncGetDocuments,
  PopupPanel,
};

describe('field-file', () => {
  beforeEach(() => {
    onChange.mockClear();
    asyncGetDocuments.mockClear();
  });
  test('Renders', () => {
    render(<FieldFile {...defaultProps} />);
  });

  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        render(<Composition />);
      });
    });
  });
  describe('shallow renders', () => {
    test('Matches Snapshot', () => {
      render(<FieldFile {...defaultProps} />);
      const tree = screen.debug();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Unit tests', () => {
    describe('Single', () => {
      beforeEach(() => {
        render(<FieldFile {...defaultProps} />);
      });

      test.skip('should pass empty list to item list if there are no files', async () => {
        // await act(async () => {
        //   component.setProps({ value: null });
        //   component.update();
        //   await new Promise((resolve) => setTimeout(resolve));
        // });
        // const itemList = component.find(ItemList);
        // // TODO: For some reason this is failing but correct in debug
        // expect(itemList.props().items).toEqual([]);
      });
      test('should call updatedate with null when we delete the file', async () => {
        const fileDeleteBtn = screen.getByRole('button', { name: /Remove/ });
        // const fileDeleteFn = component
        //   .find(ItemList)
        //   .props()
        //   .overlayButtons.find((f) => f.uid === 'remove').func;
        await UserEvent.click(fileDeleteBtn);
        // act(() => {
        //   fileDeleteFn(DEMO_FILES_DATA[0].uid);
        // });
        expect(onChange).toHaveBeenCalledWith(defaultProps.uid, null);
      });
    });
    describe('Multiple', () => {
      beforeEach(() => {
        render(<FieldFile {...defaultProps} multiple />);
      });

      test.skip('should pass empty list to item list if there are no files', () => {
      // component.setProps({ value: null });
      //   component.update();
      //   const itemList = component.find(ItemList);
      //   expect(itemList.props().items).toEqual([]);
      });
      test('should call updatedata with file removed when we delete the file', async () => {
        const fileDeleteBtn = screen.getByRole('button', { name: /Remove/ });
        await UserEvent.click(fileDeleteBtn);
        // act(() => {
        //   fileDeleteFn(DEMO_FILES_DATA[0].uid);
        // });
        expect(onChange).toHaveBeenCalledWith(
          defaultProps.uid,
          DEMO_FILES_DATA.slice(1)
        );
      });
    });
  });
});
