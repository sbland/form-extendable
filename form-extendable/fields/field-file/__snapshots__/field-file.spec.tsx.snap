// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field-file shallow renders Matches Snapshot 1`] = `undefined`;
